Sources/PenguinStructures/AnyArrayBuffer.swift

```diff
@@ -28,6 +28,51 @@ public struct AnyArrayBuffer<Storage: AnyArrayStorage> {
 
   /// The type of element stored here.
   public var elementType: Any.Type { storage.elementType }
+
+  /// Returns the result of calling `body` on the elements of `self`.
+  public func withUnsafeRawBaseAddress<R>(
+    body: (UnsafeRawPointer)->R
+  ) -> R {
+    storage.withUnsafeMutableRawBufferPointer { b in
+      body(b.baseAddress.map { .init($0) } ?? UnsafeRawPointer(bitPattern: -1).unsafelyUnwrapped)
+    }
+  }
+
+  /// Returns the result of calling `body` on the elements of `self`.
+  public mutating func withUnsafeMutableRawBaseAddress<R>(
+    _ body: (inout UnsafeMutableRawPointer)->R
+  ) -> R {
+    withMutableStorage { s in
+      s.withUnsafeMutableRawBufferPointer { b in
+        var ba = b.baseAddress ?? UnsafeMutableRawPointer(bitPattern: -1).unsafelyUnwrapped
+        return body(&ba)
+      }
+    }
+  }
+
+  /// Returns the result of calling `body` on the storage of `self`.
+  ///
+  /// `body` must not mutate elements in the storage.
+  public func withStorage<R>(_ body: (Storage) -> R) -> R {
+    body(storage)
+  }
+
+  /// Returns the result of calling `body` on the storage of `self`.
+  public mutating func withMutableStorage<R>(_ body: (Storage) -> R) -> R {
```

Comment:
I'm skeptical of exposing this because it enshrines the idea that when mutating the storage and the body isn't uniquely referenced, you want to start by copying the storage.  But that isn't true for lots of operations, e.g. inserting (a collection of) elements.

Reply:
So you think I should expose `storage` as a public get/set property, and make the extensions responsible for appropriately copying it when they want to do a mutating operation? That seems reasonable to me.

Reply:
Though it would be very nice to have a method that does the copying for the case where you do want a copy, because a lot of the mutating extensions in SwiftFusion (vector add, vector scalar multiply, differentiable move) do want a copy.

Maybe `ensureStorageUniquelyReferenced()`?

Reply:
Updated to expose `storage` and a `ensureUniqueStorage()` method.

Reply:
> mutating extensions in SwiftFusion (vector add, vector scalar multiply, differentiable move) do want a copy.

Unless I'm mistaken, they don't. `+=` wants to initialize the new vectors with the new scalar values, not write into the new vector twice, once with the old scalar values, and then again with the sums.

Reply:
Oh I see, that does sound like it would be faster than copying and then `+=`'ing the copy.

That sounds like more work to implement, and `ensureUniqueStorage()` is good enough for now, so I'll keep `ensureUniqueStorage()` but add TODOs where we use it.